Eclipse Equinox Launcher is failing on MacOS

I tried this on a java project. On. my shell i can call "/usr/libexec/java_home" without a problem.


> Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=2, No such file or directory\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=2, No such file or directory\n\tat java.base/java.lang.ProcessBuilder.start(Unknown Source)\n\tat java.base/java.lang.ProcessBuilder.start(Unknown Source)\n\tat java.base/java.lang.Runtime.exec(Unknown Source)\n\tat java.base/java.lang.Runtime.exec(Unknown Source)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:967)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:918)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:896)\n\tat org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:100)\n\tat org.eclipse.jdt.internal.launching.DetectVMInstallationsJob.run(DetectVMInstallationsJob.java:67)\n\tat org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)\nCaused by: java.io.IOException: error=2, No such file or directory\n\tat java.base/java.lang.ProcessImpl.forkAndExec(Native Method)\n\tat java.base/java.lang.ProcessImpl.<init>(Unknown Source)\n\tat java.base/java.lang.ProcessImpl.start(Unknown Source)\n\t... 10 more\n"}



I could track it to equinox [launcher code](https://github.com/eclipse-equinox/equinox/blob/188951bafc61e4e1caa95f34c203a98f398b445b/features/org.eclipse.equinox.executable.feature/library/cocoa/eclipseCocoa.c#L372), but I can't tell why it would be failing.

I don't think it is a serena issue, but I am hoping that I can see if anyone else is also having it.

- I have the same issue ，I also thought it was a permissions issue at first.

- But  I've modified the permissions. : "sudo chmod +x /usr/libexec/java_home"; "ls -l /usr/libexec/java_home
-rwxr-xr-x  37 root  wheel  135664 May  4 13:39 /usr/libexec/java_home".

- system: MacOs

- Now, Serena still reports an error：
starting EclipseJDTLS server process
INFO  2025-07-15 15:44:47,235 [SerenaAgentExecutor_0] solidlsp.ls_handler:start:157 - Starting language server process via command: /Users/louis/Documents/idea_workspace/github/oraios/serena/src/solidlsp/language_servers/static/EclipseJDTLS/vscode-java/extension/jre/21.0.7-macosx-aarch64/bin/java --add-modules=ALL-SYSTEM --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/sun.nio.fs=ALL-UNNAMED -Declipse.application=org.eclipse.jdt.ls.core.id1 -Dosgi.bundles.defaultStartLevel=4 -Declipse.product=org.eclipse.jdt.ls.core.product -Djava.import.generatesMetadataFilesAtProjectRoot=false -Dfile.encoding=utf8 -noverify -XX:+UseParallelGC -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Djava.lsp.joinOnCompletion=true -Xmx3G -Xms100m -Xlog:disable -Dlog.level=ALL -Djava.home.ignore=true -Djdt.ls.java.home=/Users/louis/Documents/idea_workspace/github/oraios/serena/src/solidlsp/language_servers/static/EclipseJDTLS/vscode-java/extension/jre/21.0.7-macosx-aarch64 -Djdt.ls.vm.location=/Users/louis/Documents/idea_workspace/github/oraios/serena/src/solidlsp/language_servers/static/EclipseJDTLS/vscode-java/extension/jre/21.0.7-macosx-aarch64 "-javaagent:/Users/louis/Documents/idea_workspace/github/oraios/serena/src/solidlsp/language_servers/static/EclipseJDTLS/vscode-java/extension/lombok/lombok-1.18.36.jar" "-Djdt.core.sharedIndexLocation=/Users/louis/.multilspy/global_cache/lsp/EclipseJDTLS/sharedIndex" -jar "/Users/louis/Documents/idea_workspace/github/oraios/serena/src/solidlsp/language_servers/static/EclipseJDTLS/vscode-java/extension/server/plugins/org.eclipse.equinox.launcher_1.7.0.v20250424-1814.jar" -configuration "/Users/louis/.multilspy/lsp/EclipseJDTLS/workspaces/87baf15efc4e4fc78b32cdf5780f697b/config_path" -data "/Users/louis/.multilspy/lsp/EclipseJDTLS/workspaces/87baf15efc4e4fc78b32cdf5780f697b/data_dir"
INFO  2025-07-15 15:44:47,240 [SerenaAgentExecutor_0] solidlsp:_start_server:712 - Sending initialize request from LSP client to LSP server and awaiting response1
INFO  2025-07-15 15:44:48,597 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 3, "message": "2025年7月15日 15:44:48 class org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin is started"}
INFO  2025-07-15 15:44:48,597 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 3, "message": "2025年7月15日 15:44:48 Started org.eclipse.buildship.core 15ms"}
INFO  2025-07-15 15:44:48,598 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 3, "message": "2025年7月15日 15:44:48 Started org.eclipse.m2e.core 0ms"}
INFO  2025-07-15 15:44:48,598 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 3, "message": "2025年7月15日 15:44:48 Main thread is waiting"}
INFO  2025-07-15 15:44:48,615 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 3, "message": "2025年7月15日 15:44:48 >> initialize"}
INFO  2025-07-15 15:44:48,615 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 3, "message": "2025年7月15日 15:44:48 Initializing Java Language Server 1.47.0.202505150330"}
INFO  2025-07-15 15:44:48,924 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 3, "message": "2025年7月15日 15:44:48 ProjectRegistryRefreshJob finished 390ms"}
INFO  2025-07-15 15:44:48,959 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 1, "message": "2025年7月15日 15:44:48 Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\n\tat java.base/java.lang.ProcessBuilder.start(Unknown Source)\n\tat java.base/java.lang.ProcessBuilder.start(Unknown Source)\n\tat java.base/java.lang.Runtime.exec(Unknown Source)\n\tat java.base/java.lang.Runtime.exec(Unknown Source)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)\n\tat org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)\n\tat org.eclipse.jdt.internal.launching.macosx.MacOSXVMInstallType.detectInstallLocation(MacOSXVMInstallType.java:149)\n\tat org.eclipse.jdt.launching.JavaRuntime.detectEclipseRuntime(JavaRuntime.java:2255)\n\tat org.eclipse.jdt.launching.JavaRuntime.initializeVMs(JavaRuntime.java:3237)\n\tat org.eclipse.jdt.launching.JavaRuntime.getVMInstallTypes(JavaRuntime.java:582)\n\tat org.eclipse.jdt.ls.core.internal.JVMConfigurator.findVM(JVMConfigurator.java:265)\n\tat org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:118)\n\tat org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:105)\n\tat org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.handleInitializationOptions(BaseInitHandler.java:128)\n\tat org.eclipse.jdt.ls.core.internal.handlers.InitHandler.handleInitializationOptions(InitHandler.java:95)\n\tat org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.initialize(BaseInitHandler.java:68)\n\tat org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.initialize(JDTLanguageServer.java:286)\n\tat java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\n\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)\n\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:128)\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:271)\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:201)\n\tat org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$1(ParentProcessWatcher.java:144)\n\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:185)\n\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:97)\n\tat org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:114)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)\n\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.io.IOException: error=0, posix_spawn failed\n\tat java.base/java.lang.ProcessImpl.forkAndExec(Native Method)\n\tat java.base/java.lang.ProcessImpl.(Unknown Source)\n\tat java.base/java.lang.ProcessImpl.start(Unknown Source)\n\t... 34 more\n"}
INFO  2025-07-15 15:44:48,966 [LSP-stdout-reader] solidlsp:window_log_message:695 - LSP: window/logMessage: {"type": 1, "message": "2025年7月15日 15:44:48 Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\n\tat java.base/java.lang.ProcessBuilder.start(Unknown Source)\n\tat java.base/java.lang.ProcessBuilder.start(Unknown Source)\n\tat java.base/java.lang.Runtime.exec(Unknown Source)\n\tat java.base/java.lang.Runtime.exec(Unknown Source)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)\n\tat org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)\n\tat org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)\n\tat org.eclipse.jdt.internal.launching.DetectVMInstallationsJob.run(DetectVMInstallationsJob.java:82)\n\tat org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)\nCaused by: java.io.IOException: error=0, posix_spawn failed\n\tat java.base/java.lang.ProcessImpl.forkAndExec(Native Method)\n\tat java.base/java.lang.ProcessImpl.(Unknown Source)\n\tat java.base/java.lang.ProcessImpl.start(Unknown Source)\n\t... 10 more\n"}

I met the same issue with Serena master branch. 
```
INFO  2025-08-27 10:42:02,352 [LSP-stdout-reader] solidlsp:logging_fn:295 - LSP: server -> client: {"jsonrpc": "2.0", "method": "window/logMessage", "params": {"type": 1, "message": "Aug 27, 2025, 10:42:02\u202fAM Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed
	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
	at java.base/java.lang.Runtime.exec(Unknown Source)
	at java.base/java.lang.Runtime.exec(Unknown Source)
	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)
	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)
	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)
	at org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)
	at org.eclipse.jdt.internal.launching.macosx.MacOSXVMInstallType.detectInstallLocation(MacOSXVMInstallType.java:149)
	at org.eclipse.jdt.launching.JavaRuntime.detectEclipseRuntime(JavaRuntime.java:2255)
	at org.eclipse.jdt.launching.JavaRuntime.initializeVMs(JavaRuntime.java:3237)
	at org.eclipse.jdt.launching.JavaRuntime.getVMInstallTypes(JavaRuntime.java:582)
	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.findVM(JVMConfigurator.java:265)
	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:118)
	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:105)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.handleInitializationOptions(BaseInitHandler.java:128)
	at org.eclipse.jdt.ls.core.internal.handlers.InitHandler.handleInitializationOptions(InitHandler.java:95)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.initialize(BaseInitHandler.java:68)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.initialize(JDTLanguageServer.java:286)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:128)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:271)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:201)
	at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$1(ParentProcessWatcher.java:144)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:185)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:97)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:114)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.io.IOException: error=0, posix_spawn failed
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(Unknown Source)
	at java.base/java.lang.ProcessImpl.start(Unknown Source)
	... 34 more\n"}}
```
another stacktrace:
```
INFO  2025-08-27 10:42:02,360 [LSP-stdout-reader] solidlsp:logging_fn:295 - LSP: server -> client: {"jsonrpc": "2.0", "method": "window/logMessage", "params": {"type": 1, "message": "Aug 27, 2025, 10:42:02\u202fAM Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed
	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
	at java.base/java.lang.Runtime.exec(Unknown Source)
	at java.base/java.lang.Runtime.exec(Unknown Source)
	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)
	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)
	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)
	at org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)
	at org.eclipse.jdt.internal.launching.DetectVMInstallationsJob.run(DetectVMInstallationsJob.java:82)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)\nCaused by: java.io.IOException: error=0, posix_spawn failed
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(Unknown Source)
	at java.base/java.lang.ProcessImpl.start(Unknown Source)
	... 10 more\n"}}
```

the permission is ok:

<img width="812" height="99" alt="Image" src="https://github.com/user-attachments/assets/dfed1d1f-69a8-48c9-a338-63092e332480" />


I searched the github and found a similar issue here: https://github.com/processing/processing4/issues/1044
It seems like an issue that not  appropriately setting up the permissions in the JDK?

> I met the same issue with Serena master branch.
> 
> ```
> INFO  2025-08-27 10:42:02,352 [LSP-stdout-reader] solidlsp:logging_fn:295 - LSP: server -> client: {"jsonrpc": "2.0", "method": "window/logMessage", "params": {"type": 1, "message": "Aug 27, 2025, 10:42:02\u202fAM Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed
> 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> 	at java.base/java.lang.Runtime.exec(Unknown Source)
> 	at java.base/java.lang.Runtime.exec(Unknown Source)
> 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)
> 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)
> 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)
> 	at org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)
> 	at org.eclipse.jdt.internal.launching.macosx.MacOSXVMInstallType.detectInstallLocation(MacOSXVMInstallType.java:149)
> 	at org.eclipse.jdt.launching.JavaRuntime.detectEclipseRuntime(JavaRuntime.java:2255)
> 	at org.eclipse.jdt.launching.JavaRuntime.initializeVMs(JavaRuntime.java:3237)
> 	at org.eclipse.jdt.launching.JavaRuntime.getVMInstallTypes(JavaRuntime.java:582)
> 	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.findVM(JVMConfigurator.java:265)
> 	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:118)
> 	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:105)
> 	at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.handleInitializationOptions(BaseInitHandler.java:128)
> 	at org.eclipse.jdt.ls.core.internal.handlers.InitHandler.handleInitializationOptions(InitHandler.java:95)
> 	at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.initialize(BaseInitHandler.java:68)
> 	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.initialize(JDTLanguageServer.java:286)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
> 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
> 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)
> 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:128)
> 	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:271)
> 	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:201)
> 	at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$1(ParentProcessWatcher.java:144)
> 	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:185)
> 	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:97)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:114)
> 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
> 	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
> 	at java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.io.IOException: error=0, posix_spawn failed
> 	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
> 	at java.base/java.lang.ProcessImpl.<init>(Unknown Source)
> 	at java.base/java.lang.ProcessImpl.start(Unknown Source)
> 	... 34 more\n"}}
> ```
> 
> another stacktrace:
> 
> ```
> INFO  2025-08-27 10:42:02,360 [LSP-stdout-reader] solidlsp:logging_fn:295 - LSP: server -> client: {"jsonrpc": "2.0", "method": "window/logMessage", "params": {"type": 1, "message": "Aug 27, 2025, 10:42:02\u202fAM Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed
> 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> 	at java.base/java.lang.Runtime.exec(Unknown Source)
> 	at java.base/java.lang.Runtime.exec(Unknown Source)
> 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)
> 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)
> 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)
> 	at org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)
> 	at org.eclipse.jdt.internal.launching.DetectVMInstallationsJob.run(DetectVMInstallationsJob.java:82)
> 	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)\nCaused by: java.io.IOException: error=0, posix_spawn failed
> 	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
> 	at java.base/java.lang.ProcessImpl.<init>(Unknown Source)
> 	at java.base/java.lang.ProcessImpl.start(Unknown Source)
> 	... 10 more\n"}}
> ```
> 
> the permission is ok:
> 
> <img alt="Image" width="812" height="99" src="https://private-user-images.githubusercontent.com/6805785/482458563-dfed1d1f-69a8-48c9-a338-63092e332480.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYyNjUyNTIsIm5iZiI6MTc1NjI2NDk1MiwicGF0aCI6Ii82ODA1Nzg1LzQ4MjQ1ODU2My1kZmVkMWQxZi02OWE4LTQ4YzktYTMzOC02MzA5MmUzMzI0ODAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjdUMDMyMjMyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVmYWFiNmYwMWI4YjI3ZTBmY2M2MDQwMTIyYzc5ZDU4Y2NhYjMwOWNhMmMxYThiZjRhNDFkNjdiNWY1MTk5MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fkIDZ30enGpCVutf2eIbreyWojJV71RrVHHh8W71KFA">
> I searched the github and found a similar issue here: [processing/processing4#1044](https://github.com/processing/processing4/issues/1044) It seems like an issue that not appropriately setting up the permissions in the JDK?

After change the permission of java lib,  the exceptions disappeared.
```
chmod -R 755  /Users/<xxx>/.serena/language_servers/static/EclipseJDTLS/vscode-java/extension/jre/21.0.7-macosx-aarch64/lib
```

> > I met the same issue with Serena master branch.
> > ```
> > INFO  2025-08-27 10:42:02,352 [LSP-stdout-reader] solidlsp:logging_fn:295 - LSP: server -> client: {"jsonrpc": "2.0", "method": "window/logMessage", "params": {"type": 1, "message": "Aug 27, 2025, 10:42:02\u202fAM Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed
> > 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> > 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> > 	at java.base/java.lang.Runtime.exec(Unknown Source)
> > 	at java.base/java.lang.Runtime.exec(Unknown Source)
> > 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)
> > 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)
> > 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)
> > 	at org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)
> > 	at org.eclipse.jdt.internal.launching.macosx.MacOSXVMInstallType.detectInstallLocation(MacOSXVMInstallType.java:149)
> > 	at org.eclipse.jdt.launching.JavaRuntime.detectEclipseRuntime(JavaRuntime.java:2255)
> > 	at org.eclipse.jdt.launching.JavaRuntime.initializeVMs(JavaRuntime.java:3237)
> > 	at org.eclipse.jdt.launching.JavaRuntime.getVMInstallTypes(JavaRuntime.java:582)
> > 	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.findVM(JVMConfigurator.java:265)
> > 	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:118)
> > 	at org.eclipse.jdt.ls.core.internal.JVMConfigurator.configureJVMs(JVMConfigurator.java:105)
> > 	at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.handleInitializationOptions(BaseInitHandler.java:128)
> > 	at org.eclipse.jdt.ls.core.internal.handlers.InitHandler.handleInitializationOptions(InitHandler.java:95)
> > 	at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.initialize(BaseInitHandler.java:68)
> > 	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.initialize(JDTLanguageServer.java:286)
> > 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
> > 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
> > 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)
> > 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:128)
> > 	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:271)
> > 	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:201)
> > 	at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$1(ParentProcessWatcher.java:144)
> > 	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:185)
> > 	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:97)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:114)
> > 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
> > 	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
> > 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
> > 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
> > 	at java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.io.IOException: error=0, posix_spawn failed
> > 	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
> > 	at java.base/java.lang.ProcessImpl.<init>(Unknown Source)
> > 	at java.base/java.lang.ProcessImpl.start(Unknown Source)
> > 	... 34 more\n"}}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > another stacktrace:
> > ```
> > INFO  2025-08-27 10:42:02,360 [LSP-stdout-reader] solidlsp:logging_fn:295 - LSP: server -> client: {"jsonrpc": "2.0", "method": "window/logMessage", "params": {"type": 1, "message": "Aug 27, 2025, 10:42:02\u202fAM Exception occurred executing command line.\nCannot run program "/usr/libexec/java_home": error=0, posix_spawn failed\njava.io.IOException: Cannot run program "/usr/libexec/java_home": error=0, posix_spawn failed
> > 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> > 	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
> > 	at java.base/java.lang.Runtime.exec(Unknown Source)
> > 	at java.base/java.lang.Runtime.exec(Unknown Source)
> > 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:1042)
> > 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:981)
> > 	at org.eclipse.debug.core.DebugPlugin.exec(DebugPlugin.java:959)
> > 	at org.eclipse.jdt.internal.launching.MacInstalledJREs.getInstalledJREs(MacInstalledJREs.java:99)
> > 	at org.eclipse.jdt.internal.launching.DetectVMInstallationsJob.run(DetectVMInstallationsJob.java:82)
> > 	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)\nCaused by: java.io.IOException: error=0, posix_spawn failed
> > 	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
> > 	at java.base/java.lang.ProcessImpl.<init>(Unknown Source)
> > 	at java.base/java.lang.ProcessImpl.start(Unknown Source)
> > 	... 10 more\n"}}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > the permission is ok:
> > <img alt="Image" width="812" height="99" src="https://private-user-images.githubusercontent.com/6805785/482458563-dfed1d1f-69a8-48c9-a338-63092e332480.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYyNjUyNTIsIm5iZiI6MTc1NjI2NDk1MiwicGF0aCI6Ii82ODA1Nzg1LzQ4MjQ1ODU2My1kZmVkMWQxZi02OWE4LTQ4YzktYTMzOC02MzA5MmUzMzI0ODAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjdUMDMyMjMyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVmYWFiNmYwMWI4YjI3ZTBmY2M2MDQwMTIyYzc5ZDU4Y2NhYjMwOWNhMmMxYThiZjRhNDFkNjdiNWY1MTk5MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fkIDZ30enGpCVutf2eIbreyWojJV71RrVHHh8W71KFA">
> > I searched the github and found a similar issue here: [[processing/processing4#1044](https://github.com/processing/processing4/issues/1044)](https://github.com/[processing/processing4/issues/1044](https://github.com/processing/processing4/issues/1044)) It seems like an issue that not appropriately setting up the permissions in the JDK?
> 
> After change the permission of java lib, the exceptions disappeared.
> 
> ```
> chmod -R 755  /Users/<xxx>/.serena/language_servers/static/EclipseJDTLS/vscode-java/extension/jre/21.0.7-macosx-aarch64/lib
> ```

The root cause is：

<img width="850" height="714" alt="Image" src="https://github.com/user-attachments/assets/7ef55a1d-dc51-4eef-80b7-76fedbe2deda" />

# Reproduce
The original permission is  -rw-r--r--

<img width="1463" height="63" alt="Image" src="https://github.com/user-attachments/assets/24999bc1-5ff1-4acc-8a6d-6c61fbc78cfc" />

and cause the exception
<img width="2556" height="389" alt="Image" src="https://github.com/user-attachments/assets/36c87f2a-ce1a-4106-b5cb-4d6aae389858" />

# chmod +x jspawnhelper

<img width="1414" height="64" alt="Image" src="https://github.com/user-attachments/assets/d50422fa-343d-4b32-b8a5-2ed35fcc075d" />

No Exception 

<img width="958" height="224" alt="Image" src="https://github.com/user-attachments/assets/04fc7799-49e8-40c0-9d60-e25225d7aa3d" />

--- a/src/solidlsp/ls_utils.py
+++ b/src/solidlsp/ls_utils.py
@@ -210,12 +210,12 @@ class FileUtils:
             if archive_type in ["tar", "gztar", "bztar", "xztar"]:
                 os.makedirs(target_path, exist_ok=True)
                 shutil.unpack_archive(tmp_file_name, target_path, archive_type)
-            elif archive_type == "zip": 
+            elif archive_type == "zip":
                 os.makedirs(target_path, exist_ok=True)
                 with zipfile.ZipFile(tmp_file_name, "r") as zip_ref:
                     for zip_info in zip_ref.infolist():
                         extracted_path = zip_ref.extract(zip_info, target_path)
-                        ZIP_SYSTEM_UNIX = 3 # zip file created on Unix system
+                        ZIP_SYSTEM_UNIX = 3  # zip file created on Unix system
                         if zip_info.create_system != ZIP_SYSTEM_UNIX:
                             continue
                         # extractall() does not preserve permissions

